photon-api/src/main/scala/com/linkedin/photon/ml/data/RandomEffectDataset.scala

```diff
@@ -406,29 +406,17 @@ object RandomEffectDataset {
    *
    * @param keyedGameDataset The data for the given feature shard, keyed by the [[REId]]s for the given [[REType]]
    * @param projectors An [[RDD]] of per-entity [[LinearSubspaceProjector]] objects
-   * @param randomEffectPartitioner A specialized partitioner to co-locate all data from a single entity, while keeping
-   *                                the data distribution equal amongst partitions
    * @return The data for the given feature shard, keyed by the [[REId]]s for the given [[REType]], with feature vectors
    *         reduced to the smallest linear subspace possible without loss
    */
   protected[data] def generateProjectedDataset(
       keyedGameDataset: RDD[(REId, (UniqueSampleId, LabeledPoint))],
-      projectors: RDD[(REId, LinearSubspaceProjector)],
-      randomEffectPartitioner: RandomEffectDatasetPartitioner): RDD[(REId, (UniqueSampleId, LabeledPoint))] =
+      projectors: RDD[(REId, LinearSubspaceProjector)]): RDD[(REId, (UniqueSampleId, LabeledPoint))] =
 
     keyedGameDataset
-      .partitionBy(randomEffectPartitioner)
-      .zipPartitions(projectors) { case (dataIt, projectorsIt) =>
-
-        val projectorLookupTable = projectorsIt.toMap
-
-        dataIt.map { case (rEID, (uID, LabeledPoint(label, features, offset, weight))) =>
-
-          val projector = projectorLookupTable(rEID)
-          val projectedFeatures = projector.projectForward(features)
-
-          (rEID, (uID, LabeledPoint(label, projectedFeatures, offset, weight)))
-        }
+      .join(projectors)
```

Comment:
Is it possible that the projectors are still causing the data to be keyed by partitions and sent to single executor to be loaded in the memory?

Reply:
I think I see what you mean - @yunboouyang and I just spoke and we're working on another solution. Yes, the same partitioner is being used even without being specified as the keyed data loses its partitioner, so Spark reuses the partitioner of the projector RDD.

Reply:
Thanks @ashelkovnykov ! Please let me know the patch and i can verify it with my dataset :)